source/blog/_posts/2015-08-04-aop.md

```diff
@@ -0,0 +1,305 @@
+
+* Advice: An advice is some code you want to add to your existing code base, e.g. to a certain model or service.
+* Pointcut: A pointcut is a point in your program, you want to connect your advice to. This could be the beginning or end
+of a function / method.
+* Aspect: A certain advice at a certain pointcut is considered an `aspect`. It is the concrete extension you do, in order to
+add e.g. logging to parts of your application.
+
+# AOP and PHP
+As AOP is more a programming paradigm then a design pattern, the language should offer some kind of support for it. Gregor Kiczales,
+as one of the authors of the AOP approach, is also one of the authors of [AspectJ](https://en.wikipedia.org/wiki/AspectJ),
+which is the most recognized AOP framework for Java.
+
+PHP does not support AOP by default, for that reason AOP is not used in real-world projects much. There are some libraries
+and ways, however, to make use of AOP in PHP projects.
+
+* [Go! AOP PHP](https://github.com/goaop/framework): Userland AOP implementation with quite some binding to e.g.
+ [Yii2](https://github.com/lisachenko/yii2-aspect), [ZF2](https://github.com/lisachenko/zf2-aspect), [Symfony](https://github.com/lisachenko/symfony-aspect)
+ and [Laravel](https://github.com/lisachenko/laravel-aspect).
+* [Exar Framework](https://github.com/techdev-solutions/exar-framework) Another userland AOP implementation for PHP
+* [appserver.io](http://appserver.io/get-started/documentation/1.0/aop.html) The PHP appserver `appserver.io` also comes
+with support for AOP
+
+AOP as a userland implementation (thus: not changing the actual PHP core or adding a PHP extension) will usually be archived
+using a [proxy pattern](https://en.wikipedia.org/wiki/Proxy_pattern). So in order to add arbitrary code to your existing
+code base, the AOP framework will need to create proxy classes for your original classes - and will provide generic join points
+at this level. In your application you will then not work with your original classes, but with the proxy objects, which are
+transparent to the original class. This is pretty much, how the [Shopware hook system](https://developers.shopware.com/blog/2015/06/09/understanding-the-shopware-hook-system/)
+works.
+
+# Go! AOP PHP and Shopware
+In order to provide a real word example, we will try to make use of the Go! AOP PHP framework within Shopware. Of course
+this should be considered a **proof of concept**, neither should this be used in production nor there is any intend to include
+such an AOP framework in Shopware in the future.
+
+## Setting it up
+First of all we set up a fresh Shopware 5.1 from Github using the CLI tools:
+
+```
+sw install:vcs -b5.1 -daop -iaop -paop
+```
+This will create Shopware inside a folder named `aop` with a new database also called `aop`. After a quick test, that
+Shopware has been set up properly, we will add the Go! AOP PHP library dependency to Shopware's `composer.json`:
+
+```
+{
```

Comment:
You can skip editing the `composer.json` and boil the installation instructions down to `composer require goaop/framework`